File: order.service/src/main/java/br/com/edu/order/infra/rest/v1/OrderController.java
package br.com.edu.order.infra.rest.v1;

import br.com.edu.order.application.usecases.CreateOrderUseCase;
import br.com.edu.order.application.usecases.DetailOrdersUseCase;
import br.com.edu.order.application.usecases.ListOrdersUseCase;
import br.com.edu.order.domain.Order;
import br.com.edu.order.domain.errors.OrderNotFoundException;
import br.com.edu.order.infra.rest.v1.request.OrderRequest;
import br.com.edu.order.infra.rest.v1.response.ListResponse;
import br.com.edu.order.infra.rest.v1.response.OrderResponse;
import br.com.edu.order.infra.rest.v1.response.Page;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.net.URI;

@RestController
@RequestMapping("/v1/orders")
public class OrderController {

    private final CreateOrderUseCase createOrderUseCase;

    private final ListOrdersUseCase listOrdersUseCase;

    private final DetailOrdersUseCase detailOrdersUseCase;

    public OrderController(CreateOrderUseCase createOrderUseCase, ListOrdersUseCase listOrdersUseCase, DetailOrdersUseCase detailOrdersUseCase) {
        this.createOrderUseCase = createOrderUseCase;
        this.listOrdersUseCase = listOrdersUseCase;
        this.detailOrdersUseCase = detailOrdersUseCase;
    }

    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<?> create(@RequestBody OrderRequest request) {

        final var input = new Order();
        input.setCreatedBy(request.getCreatedBy());
        input.setItems(request.toModelItems());

        final var id = createOrderUseCase.execute(input);

        return ResponseEntity.created(URI.create("/v1/orders/" + id)).build();
    }

    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<?> list(@RequestParam(value = "page", required = false, defaultValue = "0") Integer page,
                                  @RequestParam(value = "size", required = false, defaultValue = "100") Integer pageSize) {
        final var result = listOrdersUseCase.execute(page, pageSize);

        final var listResponse = ListResponse.builder()
                .content(result.getContent().stream().map(OrderResponse::fromModel).toList())
                .page(new Page(result.getNumber(), result.getSize(), result.getTotalPages(), result.getTotalElements()))
                .build();

        return ResponseEntity.ok(listResponse);
    }

    @GetMapping(value = "/{order_id}", produces = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<?> detail(@PathVariable("order_id") String orderId) {
        final var result = detailOrdersUseCase.execute(orderId).orElseThrow(OrderNotFoundException::new);

        final var response = OrderResponse.fromModel(result);

        return ResponseEntity.ok(response);
    }

}

File: order.service/order-service-app/src/main/kotlin/br/com/edu/order/infra/rest/v1/OrderController.kt
package br.com.edu.order.infra.rest.v1

import br.com.edu.order.domain.Order
import br.com.edu.order.domain.errors.OrderNotFoundException
import br.com.edu.order.infra.rest.v1.request.OrderRequest
import br.com.edu.order.infra.rest.v1.response.ListResponse
import br.com.edu.order.infra.rest.v1.response.OrderResponse
import br.com.edu.order.infra.rest.v1.response.Page
import br.com.edu.order.usecases.CreateOrderUseCase
import br.com.edu.order.usecases.DeleteOrdersUseCase
import br.com.edu.order.usecases.DetailOrderUseCase
import br.com.edu.order.usecases.ListOrdersUseCase
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.net.URI

@RestController
@RequestMapping("/v1/orders")
class OrderController(
    private val createOrderUseCase: CreateOrderUseCase,
    private val listOrdersUseCase: ListOrdersUseCase,
    private val detailOrderUseCase: DetailOrderUseCase,
    private val deleteOrdersUseCase: DeleteOrdersUseCase
) {

    @PostMapping(
        consumes = [MediaType.APPLICATION_JSON_VALUE],
        produces = [MediaType.APPLICATION_JSON_VALUE])
    fun create(@RequestBody request: OrderRequest): ResponseEntity<*> {
        val input = Order()
        input.createdBy = request.createdBy
        input.items = request.toModelItems()

        val id = createOrderUseCase.execute(input)

        return ResponseEntity.created(URI.create("/v1/orders/$id")).build<Any>()
    }

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE])
    fun list(
        @RequestParam(value = "page", required = false, defaultValue = "0") page: Int?,
        @RequestParam(value = "size", required = false, defaultValue = "100") pageSize: Int?
    ): ResponseEntity<*> {

        val result = listOrdersUseCase.execute(
            page!!,
            pageSize!!
        )

        val listResponse = ListResponse(
            content = result.content.stream().map(OrderResponse::fromModel).toList(),
            page = Page(page = result.number,
                pageSize = result.size,
                totalPages = result.totalPages,
                totalElements = result.totalElements
            )
        )

        return ResponseEntity.ok<Any>(listResponse)
    }

    @GetMapping(value = ["/{order_id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun detail(@PathVariable("order_id", required = true) orderId: String?): ResponseEntity<*> {
        val result = detailOrderUseCase.execute(orderId).orElseThrow { OrderNotFoundException() }

        val response = OrderResponse.fromModel(result)

        return ResponseEntity.ok<Any>(response)
    }

    @DeleteMapping
    fun deleteAll(): ResponseEntity<*> {
        deleteOrdersUseCase.execute()

        return ResponseEntity.noContent().build<Any>()
    }
}
